metadata/src/main/java/org/apache/kafka/controller/DelegationTokenControlManager.java

```diff
@@ -0,0 +1,359 @@
+                .setIssueTimestampMs(now)
+                .setExpiryTimestampMs(expiryTimestamp)
+                .setMaxTimestampMs(maxTimestamp)
+                .setTokenId(tokenId)
+                .setHmac(hmac);
+
+        records.add(new ApiMessageAndVersion(newDelegationTokenData.toRecord(), (short) 0));
+        return ControllerResult.atomicOf(records, responseData);
+    }
+
+    public ControllerResult<RenewDelegationTokenResponseData> renewDelegationToken(
+        ControllerRequestContext context,
+        RenewDelegationTokenRequestData requestData,
+        MetadataVersion metadataVersion
+    ) {
+        long now = time.milliseconds();
+
+        List<ApiMessageAndVersion> records = new ArrayList<>();
+        RenewDelegationTokenResponseData responseData = new RenewDelegationTokenResponseData();
+
+        TokenInformation myTokenInformation = getToken(requestData.hmac());
+
+        if (myTokenInformation == null) {
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(DELEGATION_TOKEN_NOT_FOUND.code()));
+        }
+
+        if (myTokenInformation.maxTimestamp() < now || myTokenInformation.expiryTimestamp() < now) {
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(DELEGATION_TOKEN_EXPIRED.code()));
+        }
+
+        if (!allowedToRenew(myTokenInformation, context.principal())) {
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(DELEGATION_TOKEN_OWNER_MISMATCH.code()));
+        }
+
+        long renewLifeTime = tokenDefaultRenewLifetime;
+        if (requestData.renewPeriodMs() > 0) {
+            renewLifeTime = Math.min(renewLifeTime, requestData.renewPeriodMs());
+        }
+        long renewTimeStamp = now + renewLifeTime;
+        long expiryTimestamp = Math.min(myTokenInformation.maxTimestamp(), renewTimeStamp);
+
+        myTokenInformation.setExpiryTimestamp(expiryTimestamp);
+
+        DelegationTokenData newDelegationTokenData = new DelegationTokenData(myTokenInformation);
+
+        responseData
+            .setErrorCode(NONE.code())
+            .setExpiryTimestampMs(expiryTimestamp);
+
+        records.add(new ApiMessageAndVersion(newDelegationTokenData.toRecord(), (short) 0));
+        return ControllerResult.atomicOf(records, responseData);
+    }
+
+    public ControllerResult<ExpireDelegationTokenResponseData> expireDelegationToken(
+        ControllerRequestContext context,
+        ExpireDelegationTokenRequestData requestData,
+        MetadataVersion metadataVersion
+    ) {
+        long now = time.milliseconds();
+
+        List<ApiMessageAndVersion> records = new ArrayList<>();
+        ExpireDelegationTokenResponseData responseData = new ExpireDelegationTokenResponseData();
+
+        if (secretKeyString == null) {
+            // DelegationTokens are not enabled
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(DELEGATION_TOKEN_AUTH_DISABLED.code()));
+        }
```

Comment:
Move to top?  Also need to check for support in metadata version?

Reply:
I added checks for metadata version in all the calls and addresses this.